File: apps/admin/src/api/searchSettings.ts
import { api } from './client';

export interface RelevanceWeights {
  title: number;
  body: number;
  tags: number;
  author: number;
}

export interface RelevanceBoosts {
  freshness: { half_life_days: number };
  popularity: { weight: number };
}

export interface RelevanceQueryParams {
  fuzziness: string;
  min_should_match: string;
  phrase_slop: number;
  tie_breaker?: number | null;
}

export interface RelevancePayload {
  weights: RelevanceWeights;
  boosts: RelevanceBoosts;
  query: RelevanceQueryParams;
}

export interface RelevanceGetOut {
  version: number;
  payload: RelevancePayload;
  updated_at: string;
}

export interface DryRunDiffItem {
  query: string;
  topBefore: string[];
  topAfter: string[];
  moved: { id: string; from: number; to: number }[];
}

export interface RelevanceDryRunOut {
  diff: DryRunDiffItem[];
  warnings: string[];
}

export async function getRelevance(): Promise<RelevanceGetOut> {
  const res = await api.get<RelevanceGetOut>('/admin/search/relevance');
  return res.data!;
}

export async function dryRunRelevance(
  payload: RelevancePayload,
  sample: string[],
): Promise<RelevanceDryRunOut> {
  const res = await api.put<
    { payload: RelevancePayload; dryRun: boolean; sample: string[] },
    RelevanceDryRunOut
  >('/admin/search/relevance', {
    payload,
    dryRun: true,
    sample,
  });
  return res.data!;
}

export async function applyRelevance(
  payload: RelevancePayload,
  comment?: string,
): Promise<RelevanceGetOut> {
  const res = await api.put<
    { payload: RelevancePayload; dryRun: boolean; comment?: string },
    RelevanceGetOut
  >('/admin/search/relevance', {
    payload,
    dryRun: false,
    comment,
  });
  return res.data!;
}

File: apps/backend/app/domains/search/schemas/search_settings.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class RelevanceWeights(BaseModel):
    title: float = 3.0
    body: float = 1.0
    tags: float = 1.5
    author: float = 0.5


class RelevanceBoosts(BaseModel):
    freshness: dict = Field(default_factory=lambda: {"half_life_days": 14})
    popularity: dict = Field(default_factory=lambda: {"weight": 1.0})


class RelevanceQueryParams(BaseModel):
    fuzziness: str = "AUTO"
    min_should_match: str = "2<75%"
    phrase_slop: int = 0
    tie_breaker: float | None = None


class RelevancePayload(BaseModel):
    weights: RelevanceWeights = Field(default_factory=RelevanceWeights)
    boosts: RelevanceBoosts = Field(default_factory=RelevanceBoosts)
    query: RelevanceQueryParams = Field(default_factory=RelevanceQueryParams)


class RelevanceGetOut(BaseModel):
    version: int
    payload: RelevancePayload
    updated_at: datetime


class DryRunDiffItem(BaseModel):
    query: str
    topBefore: list[str] = Field(default_factory=list)
    topAfter: list[str] = Field(default_factory=list)
    moved: list[dict[str, int]] = Field(default_factory=list)


class RelevancePutIn(BaseModel):
    payload: RelevancePayload
    dryRun: bool = False
    sample: list[str] = Field(default_factory=list)
    comment: str | None = None


class RelevanceApplyOut(BaseModel):
    version: int
    payload: RelevancePayload
    updated_at: datetime


class RelevanceDryRunOut(BaseModel):
    diff: list[DryRunDiffItem]
    warnings: list[str] = Field(default_factory=list)

